### ISSUE TITLE
Improve Javascript in oxygen

### ISSUE THREAD

**Issue by alriksson** [soflyy/oxygen-bugs-and-features#762]
**Describe the feature you'd like to see included in Oxygen.**

### Oxygen performance update V3.0! 🎉
Improvements of javascript features, reduce javascript usage and optimize.

**Improve the code and make the code smarter. Javascript is heavy for mobile devices on 3G network. Reduce size and optimize the code.**
1. Consider using vanilla javascript over jQuery for better performance?
2. Do not load duplicate jQuery scripts for modal and gallery and other elements adding duplicate code for each element. Can be grouped to load and have less payload on rendering.
3. Add js to a static file instead of inline. And load it template based where it is used.
3. Support for defer javascript both when all static files are deferred and loaded in the footer as well as main jQuery script. Preferably make oxygen independent from jQuery.

**What are the use cases for this feature?**
Improve performance and support with performance plugins and features. We do not want to be forced to have jQuery or any javascript as render blockers.

**Other relevant performance tickets:**
https://github.com/soflyy/oxygen-bugs-and-features/issues/755
https://github.com/soflyy/oxygen-bugs-and-features/issues/18
https://github.com/soflyy/oxygen-bugs-and-features/issues/761

**Comment by alriksson** [comment]
Consider using CSS3 over javascript libraries.
`aos.js` and `photoswipe.js` could be replaced with better libraries for performance but for the real performance, we should replace it with CSS. Javascript is heavy for mobile networks.

Most things can be done with CSS, utilize it.

**Comment by alriksson** [comment]
Libraries like jQuery are slow by design, and add a lot of additional weight to a webpage. Writing vanilla javascript will speed the website up considerably.

**Comment by louisreingold** [comment]
> Consider using CSS3 over javascript libraries. aos.js 

I'm interested. Show me some CSS only scroll animations.

**Comment by alriksson** [comment]
@louisreingold I was wrongly mentioning the aos.js lib in this thread to be replaced by CSS. But I still think and library update from 1.2 to Build 2.3.4 (soon 3.0) would be a nice improvement.
https://github.com/michalsnik/aos

**Comment by shoelaced** [comment]
On the topic of Rocket Loader:

https://support.cloudflare.com/hc/en-us/articles/200169436-How-can-I-have-Rocket-Loader-ignore-specific-JavaScripts-

**Comment by wpsumo** [comment]
Correct me if I'm wrong but I think many things used in AOS can be swapped with keyframes and transform. An example is the Pro menu which is using jquery and could be done purely with css.

**Comment by wpsumo** [comment]
@louisreingold how about this new lib: https://animate.style/ https://github.com/animate-css/animate.css you can do extra things with JS but not needed for basic usage.

**Comment by wpsumo** [comment]
@KittenCodes @louisreingold @Spellhammer https://make.wordpress.org/themes/2021/10/04/the-performance-impact-of-using-jquery-in-wordpress-themes/

**Comment by yankiara** [comment]
+1000

**Comment by leolll** [comment]
Getting rid of jQuery would be huge for performance.  Or at least some WP hooks where community members can rewrite and replace the inline javascript with vanilla JS.  I've done this for the Accordion and Sticky Header but it feels pretty hacky without a proper hook.

**Comment by wpsumo** [comment]
Even 2021 they used jquery for the composite elements. All javascript in oxygen frontend is not many row of code. Should not take long to rewrite in vanilla js and use static js file instead of inlining. Of course inlining is easier to only output on page where it is used than compiling a static file which only contain used ones.

**Comment by leolll** [comment]
Whenever I've seen Oxygen mentioned by bloggers or YouTubers, it's always in reference to its page speed.  I really hope oxygen keeps that edge and going vanilla JS (or pure CSS) for components would be huge.  Or just help enable the community (via hooks) to replace the render-blocking jQuery with their own JS.

**Comment by yankiara** [comment]
Hi,
In the process of rewriting promenu JS and moving it to an external file, I'm struggling to remove inline JS from wp_footer...

Inline code is added in component-framework\components\classes\menu-pro-class.php with:
```
        // add JavaScript code only once and if shortcode presented
        if ($this->js_added !== true) {
            add_action( 'wp_footer', array( $this, 'output_js' ) );
            $this->js_added = true;
        }
```

... where $this refers to OxyProMenu class.

So I try to dequeue this with:
```
    add_action( 'init', function() {
        remove_action( 'wp_footer', [ 'OxyProMenu', 'output_js' ] );
    } );
```

And it doesn't work.

Yet, when I echo __CLASS__ from the pro menu class, it is 'OxyProMenu'.

What is really weird is that if I edit Oxygen's code like this:
```
        // add JavaScript code only once and if shortcode presented
        if ($this->js_added !== true) {
            add_action( 'wp_footer', array( 'TestString', 'output_js' ) );
            $this->js_added = true;
        }
```

And if I dequeue with this:
```
    add_action( 'init', function() {
        remove_action( 'wp_footer', [ 'TestString', 'output_js' ] );
    } );
```

It works perfectly!

So, I might be missing something about add_action and classes...

Anyone?

**Comment by wpsumo** [comment]
Even pro theme from theme.co removed jQuery dependencies in the latest update. It's way to slow progress on this in oxygen.

It's time to rewrite everything in vanilla JavaScript.

**Comment by leolll** [comment]
The appeal of Oxygen is fast page loads. That's not gonna happen if you use any component that loads jQuery. It's a large undertaking, but at least if we had WP hooks where we could replace the jQuery code with Vanilla JS, the community could probably handle the majority of the components. I've personally replaced the menu & toggles with Vanilla JS but without WP hooks it's brittle code.

**Comment by wpsumo** [comment]
@leolll Please share if you want, happy to convert the pro menu and toggles. I agree with you without hooks its not ideal. But as we have to wait too long for a fix its better than not doing it.

@Spellhammer to add to my suggestion above, for lightbox we can use Lity, it is now also forked and rewritten in vanilla js
https://github.com/naton/lity#other-changes

Making it a perfect alternative  now when Naton forked it to a JS version.

**Comment by lucianahanan** [comment]
+10000
if oxygen replaces all of its jquery usage with vanilla js, that would be perfect !!

**Comment by Dade88** [comment]
I would add having JS in files should also be considered, as having <script> tags around prevents from using Content Security Policy in a useful way (having CSP with unsafe-inline is like not having CSP enabled).
